rfcs/20200826-keras-nlp-scoping-design.md

```diff
@@ -0,0 +1,79 @@
+# Keras NLP
+
+| Status        | Proposed      |
+:-------------- |:---------------------------------------------------- |
+| **Author(s)** | Zhenyu Tan (tanzheny@google.com), Mark Momernick (momernick@google.com), Francois Chollet (fchollet@google.com) |
+| **Updated**   | 2020-08-26                                           |
+
+
+## Objective
+
+We aim at describing the scope of [keras-nlp](https://github.com/keras-team/keras-nlp), especially:
+- What keras-nlp should include
+- Boundaries between keras-nlp and [tensorflow addons](https://github.com/tensorflow/addons)
+- Boundaries between keras-nlp and [tensorflow model garden](https://github.com/tensorflow/models)
+- Boundaries between keras-nlp and [tf.keras](https://www.tensorflow.org/api_docs/python/tf/keras).
+- Boundaries between keras-nlp and [tf.text](https://www.tensorflow.org/tutorials/tensorflow_text/intro).
+
+## Motivation
+
+Natural language processing (nlp) has become an essential component in many machine learning solutions.
+However, a Keras-native modeling solution is still lacking for TF2.
+
+## User Benefit
+
+We hope to help machine learning engineer benefit from re-using Keras native, optimized, and well-tested
+components to build their models. We hope to help research scientists benefit from subclassing and re-composing
+Keras primitive components to test new research ideas. 
+
+## Design Proposal
+
+We propose keras-nlp to include most transformer-based modules, specifically:
+- Keras layer components such as Transformer encoder and decoder blocks.
+- Keras task components such as masked language, span labeler and named entity recognition.
+- Tensorflow operations such as beam search.
+- Keras optimizer utilities such as learning rate schedules widely used.
+- Data loader and preprocessing for different dataset, such as SQUAD, GLUE.
+
+Criteria for keras-nlp:
+- Widely accepted building components that serve various modeling tasks.
+- Tasks that improves model training.
+- Operations that will work in CPU/GPU/TPU.
+
+Boundaries between keras-nlp and TF.Text:
+- TF.Text will contain all pre-processing operations, such as WordPiece Tokenizer, n-grams, that handles strings.
+- keras-nlp will contain components after tokenization.
+
+Boundaries between keras-nlp and Tensorflow Addons:
+- Highly experimental modeling, layers, losses, etc, live in addons.
+- Components from addons will graduate to Model Garden, given it incurs more usage,
```

Comment:
I suppose we graduate on Tensorflow. Right? Or also in Keras when it will be in the standalone repository again?
We already started a discussion at https://github.com/tensorflow/community/pull/241 some months ago about if considering Keras standalone as an upstream target as Tensorflow. 
If this is the case, like with the next Keras-cv RFC, we need a little bit to disambiguate this for contributors.

Reply:
Graduating to Keras or TF is mostly a matter of "if this is a Keras object or not", right? This is hopefully clear enough

Reply:
Yes this was taken for granted what I meant is that could be confusing for contributors to start a PR in Addons or in a "satellite" Keras repository.
Also our experience is that it is hard to enforce to open a ticket before starting to work on a PR but at least with a policy of public issues before implementation we could try to have a better visibility on the third party contributors routing process.

Sometimes PR a re-routed downstream from Tensorflow to TFA. Do you will also re-route downstream from Keras standalone to keras-nlp (or keras-cv), tfa? 

Then I suppose also that here you will have some kind of "internal" development resources/efforts so I think that for these activities the process could be a little bit more under control then "sparse" user contributions that instead they can instead start in every mentioned repository. 
Having a little bit of visibility on this could help us to not spent time to review PRs that have some parallel internal potential duplicated activities.